File: 530. Minimum Absolute Difference in BST/getMinimumDifference.js
// Definition for a binary tree node.
class TreeNode {
  constructor(val = 0, left = null, right = null) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}
/**
 * @param {TreeNode} root
 * @return {number}
 */
const getMinimumDifference = (root) => {
  let prev = null;
  let minDiff = Infinity;

  const inorderTraversal = (node) => {
    if (!node) {
      return;
    }

    inorderTraversal(node.left);

    if (prev) {
      const diff = Math.abs(node.val - prev.val);
      minDiff = Math.min(minDiff, diff);
    }

    prev = node;

    inorderTraversal(node.right);
  }

  inorderTraversal(root);
  return minDiff;
};


const root = new TreeNode(4);
root.left = new TreeNode(2);
root.right = new TreeNode(6);

root.left.left = new TreeNode(1);
root.left.right = new TreeNode(3);
      
console.log('getMinimumDifference: ', getMinimumDifference(root));

File: 530. Minimum Absolute Difference in BST/getMinimumDifference.py
from typing import Optional

# Definition for a binary tree node.
class TreeNode:
  def __init__(self, val=0, left=None, right=None):
    self.val = val
    self.left = left
    self.right = right

class Solution:
  def getMinimumDifference(self, root: Optional[TreeNode]) -> int:
    prev = None
    minDiff = float('inf')

    def inorderTraversal(node):
      nonlocal minDiff, prev
      if not node:
        return
      
      inorderTraversal(node.left)

      if prev is not None:
        diff = abs(node.val - prev.val)
        minDiff = min(minDiff, diff)

      prev = node
      
      inorderTraversal(node.right)


    inorderTraversal(root)
    return minDiff
  

root = TreeNode(4)
root.left = TreeNode(2)
root.right = TreeNode(6)

root.left.left = TreeNode(1)
root.left.right = TreeNode(3)
      
solution = Solution()
print('getMinimumDifference: ', solution.getMinimumDifference(root))
